### PR TITLE
fix a leak where compiled results lived too long

### LINKED ISSUE #1600: make pmap compilation cache "act globally" with no axis name
Both the `jit` and `pmap` compilation caches are global. The `jit` cache acts that way:

```python
from jax import jit

def f(x):
  print("miss")
  return x

jit(f)(3)  # prints
jit(f)(3)  # doesn't print, i.e. cache hit
```

But the `pmap` one seems not to when you don't provide a name for the axis:

```python
from jax import pmap
from jax.lib import xla_bridge as xb
import jax.numpy as np

def f(x):
  print("miss")
  return x

x = np.ones(xb.device_count())
pmap(f)(x)  # prints
pmap(f)(x)  # prints

pmap(f, axis_name='i')(x)  # prints
pmap(f, axis_name='i')(x)  # doesn't print
```

The issue is that when an axis name isn't provided, we create a dummy one, and we create a new one each time we evaluate `pmap(f)` to produce a new pmapped version of `f`. We should probably remedy that, since patterns like

```python
for i in range(100):
  x = pmap(update)(x)
```

are handy, but they're currently a foot-gun with respect to compilation caching.

### PR DESCRIPTION
**Pull request by mattjj**
The original repro @levskaya showed us was essentially this OOM:

```python
for i in range(40):
  f = jit(lambda: 1. * np.ones((300, 1024, 1024)))
  f().block_until_ready()
```

Even though f was being rebound on every iteration, the cache entries corresponding to the previous iterations of the loop were sticking around.

Instead, if the user drops all references to a function, we want to clear the corresponding compilation cache entries (since they can never be used).

The fix here is to use a two-level cache for compiled code: the first level is a WeakKeyDictionary keyed by the raw Python callable underlying the WrappedFun, and the second level is a regular dictionary keyed by (transforms, params, args). Because this logic is now present in  linear_util.py:cache, the implementations of `WrappedFun.__eq__` and `WrappedFun.__hash__` may be superfluous now.

One unintended consequence is that this implementation now avoids using fastcache.crlu_cache for the jit and pmap compilation caches. It was easier to implement this logic in pure Python. We might want to revise this for performance reasons. ~*That means Travis might fail due to OOMs.*~ @hawkinsp points out that we have no way of knowing a priori if, say, some Google user's code is relying on the LRU behavior that this PR is removing (hopefully temporarily). If this change breaks someone in Google, I'll roll it back asap!

This change does not affect op-by-op execution; those caches use util.py:cache, which are still backed by fastcache.clru_cache. This change only affects jit and pmap caching because those are the only caches backed by linear_util.py:cache.

This PR also incidentally fixed #1600.

TODO:
- [x] run a micro-benchmark before/after, maybe the one mentioned in #1224 
- [x] address reviewer comments

As a micro-benchmark for cache hits, I used this script locally on CPU:

```python
from jax import jit
import time

N = 100_000

f = jit(lambda x: x)
f(2)  # warmup

# time with blocking
tic = time.time()
for _ in range(N):
  f(2).block_until_ready()
toc = time.time()
print((toc - tic) / N)

# time without blocking
tic = time.time()
x = 2
for _ in range(N-1):
  x = f(x)
f(x).block_until_ready()
toc = time.time()
print((toc - tic) / N)
```

```
# on master
0.00012189582347869873
4.287614107131958e-05

# on this pr
0.00011845468282699585
4.283436298370361e-05
```

(The timings are essentially identical if we instead use `f = jitlambda x: x + x)`.)

I think that's strong enough evidence that this isn't regressing our performance on the cache hit path.

### CODE REVIEW

**Review by levskaya — COMMENTED** [review]
LGTM.  We may just want to keep in mind that there are no limits on these function caches just in case we run into some strange oom issue down the road associated with an excessive amount of compilation.

File: jax/linear_util.py [hunk 1]
```diff
@@ -208,15 +196,16 @@ def wrap_init(f, params={}):
...
+  caches = weakref.WeakKeyDictionary()
+  def memoized_fun(f, *args):
+    cache = caches.setdefault(f.f, {})
+    key = (f.transforms, f.params, args)
+    if key in cache:
+      ans, stores = cache[key]
+      f.populate_stores(stores)
+    else:
+      ans = call(f, *args)
+      cache[key] = (ans, f.stores)
```
Comment (shoyer) [inline jax/linear_util.py]: I spent a little while thinking about possible race conditions with multi-threading and convinced myself that this is fine. We don't do any clean-up, so `cache[key]` is guaranteed to be valid if `key in cache`. In the worst case, code gets compiled twice.

Fun fact: `dict.setdefault` is atomic, but didn't always used to be: https://bugs.python.org/issue13521

File: jax/linear_util.py [hunk 2]
```diff
@@ -208,15 +196,16 @@ def wrap_init(f, params={}):
...
-  def cached_fun(f, *args):
-    ans, f_prev = cached_fun_body(f, args)
-    if id(f_prev) != id(f):
-      f.populate_stores(f_prev)
+def cache(call):
+  caches = weakref.WeakKeyDictionary()
+  def memoized_fun(f, *args):
+    cache = caches.setdefault(f.f, {})
+    key = (f.transforms, f.params, args)
+    if key in cache:
```
Comment (hawkinsp) [inline jax/linear_util.py]: Can we avoid the double dictionary lookup? (Call `.get()`)

File: jax/linear_util.py [hunk 3]
```diff
@@ -208,15 +196,16 @@ def wrap_init(f, params={}):
...
+      ans, stores = cache[key]
```
Comment (hawkinsp) [inline jax/linear_util.py]: I'm confused why this works at all. Doesn't key (the tuple) lose its only strong reference at the end of this function? Wouldn't this mean that all functions can and will always be evicted?

Comment (hawkinsp) [reply jax/linear_util.py]: Ok, I read this more carefully. It's fine. But the difference between `caches` and `cache` is easy to miss on a quick skim.

File: jax/linear_util.py (hunk 1 above)
Comment (mattjj) [reply jax/linear_util.py]: I'm glad you thought about this hard case, because I didn't :)

File: jax/linear_util.py (hunk 2 above)
Comment (mattjj) [reply jax/linear_util.py]: Great idea, thanks!

**Comment by shoyer** [comment]
I also misread caches/cache on first look. Let's pick a different name for
one of them.

File: jax/linear_util.py (hunk 2 above)
Comment (mattjj) [reply jax/linear_util.py]: Done.

**Comment by mattjj** [comment]
For posterity: @hawkinsp gave me the verbal LGTM, I swear.
